BinScatter/R/BinScatter.R

```diff
@@ -0,0 +1,258 @@
+                                   partialBinType, intercept) %>% reduce(cbind))
+  }
+  df_reg <- cbind(df_reg, data[, all_of(c(linpartial_var))])
+  df_reg <- df_reg %>% drop_na()
+  
+  
+  ##------------------------------------------------------------------------------------------------
+  ## Calculate partialed means (and std. err.) for `y_var` (and recenter)
+  ##------------------------------------------------------------------------------------------------
+  
+  model <- sprintf("%s ~ %s - 1", y_var, paste0(names(df_reg)[-1], collapse = " + "))
+  coef <- names(df_reg)[2:(1+nBins)]
+  y_mean <- c()
+  y_se <- c()
+  
+  if (intercept) { 
+    model <- model %>% str_sub(1,-5)
+    coef <- names(df_reg)[2:nBins]
+    y_mean <- append(y_mean, 0)
+    y_se <- append(y_se, 0)
+  }
+  reg <- lm(eval(model), df_reg)
+  reg_sum <- reg %>% summary()
+  y_mean <- append(y_mean, reg$coefficients[coef] %>% unname())
+  y_se <- append(y_se, reg_sum$coefficients[coef,"Std. Error"] %>% unname())
+  
+  ## Write regression table
+  if (!is.null(tab_path)) {
+
+    file <- sprintf("%s.txt", tab_path)
+    sink(file = file, type = "output")
+    print(reg_sum)
+    cat("Observations:", nobs(reg))
+    sink()
+  }
+
+  partialed_mean <- list(mean = y_mean, se = y_se)
+  y_vals         <- partialed_mean[["mean"]]
+  if (scale_yvar) { y_vals <- y_vals + ( mean( data[,eval(y_var)] ) - mean( y_vals ) ) }
+  
+  
+  ##------------------------------------------------------------------------------------------------
+  ## Make and save plot
+  ##------------------------------------------------------------------------------------------------
+  
+  df_plot <- cbind(y_vals, x_vals) %>% as.data.frame()
+  if (!is.null(ci)) {
+    if (!is.numeric(ci) | ci <= 0 | ci >= 1) {
+      ci <- 0.95
+      warning("ci is not numeric or outside (0,1), set to 0.95")
+    }
+    df_plot$y_low <- y_vals + qnorm( (1-ci)/2 ) * partialed_mean[["se"]]
+    df_plot$y_up  <- y_vals - qnorm( (1-ci)/2 ) * partialed_mean[["se"]]
+  }
+ 
+  if (is.null(plot_xlab)) { plot_xlab <- x_var }
+  if (is.null(plot_ylab)) { plot_ylab <- y_var }
+  
+  plot <- ggplot(df_plot, aes(x = x_vals, y = y_vals)) +
+    geom_point(size = point_size, color = point_color, shape = point_shape, fill = point_fill) + 
+    labs(x = plot_xlab, y = plot_ylab) +
+    theme_bw() +
+    {if (!is.null(ci)) 
+      geom_errorbar(aes_string(ymin = 'y_low', ymax = 'y_up'), width = bar_width, size=bar_size)} +
+    {if (!is.null(plot_xlim)) scale_x_continuous(limits = plot_xlim) } +
+    {if (!is.null(plot_ylim)) scale_y_continuous(limits = plot_ylim) } + 
+    theme(axis.title.x = element_text(size = axis_title_x_size),
+          axis.title.y = element_text(size = axis_title_y_size),
+          axis.text = element_text(size = axis_text_size))
+  
+  if (!is.null(plot_path)) {     
+    ggsave(sprintf('%s.pdf', plot_path), width = fig_width, height = fig_height, units = "in")
+  }
+  
+  return(list(reg, plot))
```

Comment:
This script is a little long. It's fairly clear so I don't think it's a big problem, but if there were two or three more functions we thought could go in their own scripts (or in a helper script), that might be nice.

Reply:
@jmshapir I took out the functions for getting midpoint/mean value, as well as making bins. You can see the result in the `Scatter.R` file in [0bc0013](https://github.com/gslab-econ/gslab_r/pull/22/commits/0bc0013ec9f9ec7675a505a5464530b3366b09a4). If it is still too long please let me know & I can take out more!